### PR TITLE
Allow show hidden metrics in kube-apiserver

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add a flag to show hidden metrics.

Refer to [stability framework kep](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md):
> On a subsequent release (when the metric's deprecatedVersion is equal to current_kubernetes_version - 1)), a deprecated metric will become a hidden metric. Unlike their deprecated counterparts, hidden metrics will no longer be automatically registered to the metrics endpoint (hence hidden). However, they can be explicitly enabled through a command line flag on the binary (i.e. '--enable-hidden-metrics=really_deprecated_metric'). This is to provide cluster admins an escape hatch to properly migrate off of a deprecated metric, if they were not able to react to the earlier deprecation warnings. Hidden metrics should be deleted after one release.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
It's a lightweight implementation against the KEP.

**Does this PR introduce a user-facing change?**:

```release-note
New flag `--show-hidden-metrics-for-version` in kube-apiserver can be used to show all hidden metrics that deprecated in the previous minor release.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md
```

### CODE REVIEW

**Review by logicalhan — COMMENTED** [review]
/cc @ehashman 

This looks reasonable to me. 

@lavalamp, does this flag look okay to you?

**Comment by ehashman** [comment]
There's no tests included in this PR, we should really include a unit test at minimum.

I'm assuming this is sort of a follow-up to #81970.

@logicalhan we previously discussed that we wanted this to be a feature flag rather than a CLI option as adding a CLI option removes the possibility of removing it/changing schema in the future, and the KEP specifies that we should be able to list specific hidden metrics to enable/disable (rather than a blanket disable for all of them). I think for now it would make more sense to go with the feature flag approach as I previously looked at?

File: cmd/kube-apiserver/app/options/options.go
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: What is to prevent an admin from:
1. in version X, turn this on to get back deprecated metric M
2. in version Y, not turn this off, and therefore fail to notice metric N is being deprecated
3. in version Z, metric N is removed with (effectively) no warning.

To prevent this scenario, I suggest forcing the user to specify the version for which they want to show metrics as the value of this flag rather than "true".

Comment (logicalhan) [reply cmd/kube-apiserver/app/options/options.go]: I like the suggestion, it seems reasonable to me.

Comment (ehashman) [reply cmd/kube-apiserver/app/options/options.go]: I think this is a good idea (and I also wouldn't mind getting rid of the KEP specification to be able to turn on/off individual metrics and just make it binary-wide for hidden) so perhaps we should update the KEP to include this?

Comment (logicalhan) [reply cmd/kube-apiserver/app/options/options.go]: Personally I think we should be able to turning off individual metrics, since we have had memory leak issues in the past. These don't tend to manifest except under certain cardinality conditions, so it's possible that metrics like that will get sneak by and get released at some point. I'm thinking of that as a beta/GA feature though. Always good to have that safety net of being able to turn off a memory leak without having to patch and release kubernetes.

**Comment by logicalhan** [comment]
> @logicalhan we previously discussed that we wanted this to be a feature flag rather than a CLI option as adding a CLI option removes the possibility of removing it/changing schema in the future, and the KEP specifies that we should be able to list specific hidden metrics to enable/disable (rather than a blanket disable for all of them). I think for now it would make more sense to go with the feature flag approach as I previously looked at?

I was thinking maybe we should just separate the functionality. A generic toggle for deprecated/hidden metrics but we should still eventually have the ability to disable metric[s] on demand. I want the safety net to be able to get rid of a metric memory leak without having to release a new version of kubernetes.

**Comment by ehashman** [comment]
So from this discussion I'm hearing we should specify two things:

- what baseline version we want base unhiding metrics against
- what specific metrics we want to unhide

Sounds like we now need two CLI flags... can we have a quick design discussion somewhere, maybe on the KEP, before moving forward with this?

/hold

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: > I like the suggestion, it seems reasonable to me.

+1 
It's a really good point!

**Comment by RainbowMango** [comment]
@ehashman 

> what specific metrics we want to unhide

If I understand correctly, here @logicalhan want to `hide` specific metrics, not `unhide`. :) 

I doubt if it is convincing enough to add a flag(hide metrics) just to prevent memory leak?

**Comment by RainbowMango** [comment]
From this discussion, personally I think we should provide a feature gate as in #81970 which can provide stability guarantees for the flag(s).

e.g
* in 1.17 flag(alpha) feature gate(default off)
* in 1.18 flag(beta) feature gate(default on)
* in 1.19 flag(stable) feature gate(removed)

**Comment by logicalhan** [comment]
> From this discussion, personally I think we should provide a feature gate as in #81970 which can provide stability guarantees for the flag(s).
> 
> e.g
> 
> * in 1.17 flag(alpha) feature gate(default off)
> * in 1.18 flag(beta) feature gate(default on)
> * in 1.19 flag(stable) feature gate(removed)

This is basically not at all the conclusion I am coming to. How are you arriving at feature gates from this conversation?

**Comment by logicalhan** [comment]
> I doubt if it is convincing enough to add a flag(hide metrics) just to prevent memory leak?

We have already had to do this: https://github.com/kubernetes/kubernetes/pull/74636. We should not require a new release of kubernetes in order to turn off a bad metric.

**Comment by RainbowMango** [comment]
> How are you arriving at feature gates from this conversation?

I like @lavalamp 's idea.  I am a little worried we will lose the possibility of improving this flag when we get another good idea. 

Am I over-cautious?

**Comment by logicalhan** [comment]
@lavalamp’s idea (as I understand it from talking to him about it today) is that the flag takes in a string parameter, a version for which we will show what would otherwise be hidden metrics. This eliminates the danger of what happens when someone forgets to untoggle a flag after migrating a set of deprecated metrics.

Being able to hide metrics is something which we can address orthogonally, using another flag.

**Comment by RainbowMango** [comment]
> @lavalamp’s idea (as I understand it from talking to him about it today) is that the flag takes in a string parameter, a version for which we will show what would otherwise be hidden metrics. This eliminates the danger of what happens when someone forgets to untoggle a flag after migrating a set of deprecated metrics.
> 
> Being able to hide metrics is something which we can address orthogonally, using another flag.

OK. Copy that.  as well as 

> We have already had to do this: #74636. We should not require a new release of kubernetes in order to turn off a bad metric.

I will push a change according @lavalamp's comments.

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [inline cmd/kube-apiserver/app/options/options.go]: Which version format should be expected?  actually, we don't care patch version, right?
- x.y.z
- x.y
- x.y.0

**Review by RainbowMango — COMMENTED** [review]
Updated according to @lavalamp 's comments. Also with questions from myself.

**Comment by ehashman** [comment]
> If I understand correctly, here @logicalhan want to `hide` specific metrics, not `unhide`. :)

This does not match my understanding. The point of this flag is to *show* (i.e. un-hide) hidden metrics. Metrics become hidden by default upon reaching a certain release version as part of the metrics stability deprecation process. This flag allows an end user to override the default hiding of metrics.

File: cmd/kube-apiserver/app/options/options.go
Comment (lavalamp) [reply cmd/kube-apiserver/app/options/options.go]: I think "vX.Y" or "x.y". No need to use the patch version (changes at the wrong time) and it's definitely misleading to always want a ptach version of 0.

**Comment by logicalhan** [comment]
> @logicalhan wants a separate (and IMO unrelated) flag, to emergency-disable a metric that has had a combinatorial explosion in distinct labels.
> 
> I think the format for taking the version should be just major and minor versions: e.g. "v1.16"

Yes, this is what I basically meant.

Sorry I may not have been explaining myself very well, most of this stuff I had commented from a phone, so my descriptions were likely too abbreviated.

File: cmd/kube-apiserver/app/options/options.go
Comment (ehashman) [reply cmd/kube-apiserver/app/options/options.go]: I'd vote `x.y` personally.

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [inline cmd/kube-apiserver/app/options/options.go]: Should we add a new flag set? e.g. "Metrics flags:". I think it makes sense.

**Review by RainbowMango — COMMENTED** [review]
Updated. (version format x.y)

New question in comments.

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: Also, invite @brancz for review.

Comment (brancz) [reply cmd/kube-apiserver/app/options/options.go]: Yes I think a new flagset is appropriate, as there will be more to come.

**Comment by RainbowMango** [comment]
1.17 code freeze is coming soon.  We need to hurry up. 
We need to add the flag to other Kubernetes binaries, they rely on this.

**Comment by RainbowMango** [comment]
@ehashman Do you think we need more test cases? 
IMO, we need some end to end tests,  integration tests, or E2E tests? I'm not sure. It's my first time deal with flags.

**Comment by ehashman** [comment]
@RainbowMango I don't think there is time for this (and updates to all the other binaries) to land in 1.17, especially with folks also busy with prep for KubeCon. We're past enhancement freeze, and we need to update the KEPs with the new format of this flag. Has that been done?

**Comment by brancz** [comment]
After reading through this discussion again and the reasoning. I think I'm ok with what we have here landing in 1.17, and following up with a new KEP for 1.18 to arbitrarily hide metrics, as that nominally doesn't have to have anything to do with their deprecation or stability state.

**Comment by logicalhan** [comment]
> After reading through this discussion again and the reasoning. I think I'm ok with what we have here landing in 1.17

Great. My intent was that the beta launch of the stability work allows for the proper deprecation of metrics

> and following up with a new KEP for 1.18 to arbitrarily hide metrics, as that nominally doesn't have to have anything to do with their deprecation or stability state.

Actually I disagree with a part of this. Yes we will need a new KEP, but it will be the 'Bring stability framework to GA` KEP. Beta allows for metric deprecation. GA allows for metrics to be marked as Stable with enforcement of stability guarantees. Being able to disable a metric is a necessary feature which I have felt we require in order to graduate a metric to being considered stable. If we cannot disable a metric when it becomes problematic, then we may end up being forced to reneg on our stability guarantees (which is highly, highly undesirable, since they would then not be guarantees). Rather, if we can tell people, for whom a metric becomes problematic that they can explicitly disable that metric, then we can (1) allow people to recover their cluster without requiring a patch (2) maintain our API stability guarantees.

**Comment by brancz** [comment]
I generally think it's orthogonal, but I can understand how it can be considered part of the metrics stability framework KEP. It's a nuance I feel, but happy with moving forward as you describe it.

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

File: cmd/kube-apiserver/app/options/options.go
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: Either add a comment or name it "ShowHiddenMetricsForVersion"?

File: cmd/kube-apiserver/app/options/options.go
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: `--show-hidden-metrics-for-version` would be slightly clearer; up to you, though?

Comment (logicalhan) [reply cmd/kube-apiserver/app/options/options.go]: That sounds reasonable to me.

Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: +1

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: Done.

File: cmd/kube-apiserver/app/options/options.go
Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: Done.

File: staging/src/k8s.io/component-base/metrics/registry.go
Comment (RainbowMango) [inline staging/src/k8s.io/component-base/metrics/registry.go]: Oh, my god.  I forgot to update this flag name in error message.
Same as below.

**Review by RainbowMango — COMMENTED** [review]
I suggest double confirm the flag description as well as the error message.
It's important because it will be exposed to every related kubernetes instance.

My proposal:
For flag description: `The previous version(x.y) for which you want to show hidden metrics. Only the previous minor version is allowed.`

For validation message:
- Given `x.y.z` :  `specified --show-hidden-metrics-for-version 'x.y.z' not follows the version format x.y`
- Given `x.y`(too small version): `specified --show-hidden-metrics-for-version 'x.y' is smaller than expected. Only the previous minor version is allowed`
- Given `x.y`(too bigger version): `specified --show-hidden-metrics-for-version 'x.y' is bigger than expected. Only the previous minor version is allowed`

**Review by logicalhan — APPROVED** [review]
/lgtm

**Comment by RainbowMango** [comment]
@lavalamp 
Any comments from you? Can we push forward now?

**Comment by ehashman** [comment]
/retitle Allow show hidden metrics in kube-apiserver
/hold cancel

With KEP merged, I am happy with this. Just need an approver from apiserver OWNERS.

Note that we will also need this flag implemented in the rest of the kube binaries, not just the apiserver (e.g. kubelet, scheduler, etc.)

**Review by ehashman — APPROVED** [review]
/lgtm

**Comment by RainbowMango** [comment]
> With KEP merged, I am happy with this. Just need an approver from apiserver OWNERS.

Let's ping @kubernetes/sig-api-machinery-pr-reviews , code freeze is fast approaching... this Thursday

**Comment by RainbowMango** [comment]
> Note that we will also need this flag implemented in the rest of the kube binaries, not just the apiserver (e.g. kubelet, scheduler, etc.)

Yeah, I know that. I intended to deal with them after this PR. Because they will share the validation logic from the component base.

File: cmd/kube-apiserver/app/options/options.go
Comment (sttts) [inline cmd/kube-apiserver/app/options/options.go]: What is this version? Which formats are accepted? What does the last sentence mean?

Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go]: Why not `--show-hidden-metrics` without any argument?

File: cmd/kube-apiserver/app/options/validation.go
Comment (sttts) [inline cmd/kube-apiserver/app/options/validation.go]: why the wrapper?

File: cmd/kube-apiserver/app/server.go
Comment (sttts) [inline cmd/kube-apiserver/app/server.go]: convention: len(*) > 0

File: cmd/kube-apiserver/app/options/options.go
Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go]: This kind of flag is incompatible by design to the previous version on each version change.

Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/options.go]: > Why not --show-hidden-metrics without any argument?

See the discussion above:
https://github.com/kubernetes/kubernetes/pull/84292#discussion_r338804880

Also updated to KEP:
[why not --show-hidden-metrics](https://github.com/kubernetes/enhancements/blob/73a9f147245bee11d6590e9c6e09c6ee349bcac7/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#why-not-bool-flag)

File: cmd/kube-apiserver/app/options/validation.go
Comment (RainbowMango) [reply cmd/kube-apiserver/app/options/validation.go]: Share validation logic with other kubernetes binaries. Is it ok?

File: cmd/kube-apiserver/app/options/options.go
Comment (sttts) [reply cmd/kube-apiserver/app/options/options.go]: Ok, got that.

Another point: I have not read the KEP. I have no clue what this flag does. What does "show" mean? What does "hidden" mean? This must be super clear from the flag description.

File: cmd/kube-apiserver/app/options/validation.go
Comment (sttts) [reply cmd/kube-apiserver/app/options/validation.go]: why not just call `metrics.ValidateShowHiddenMetricsVersion` directly below?

File: cmd/kube-apiserver/app/options/options.go
Comment (brancz) [reply cmd/kube-apiserver/app/options/options.go]: maybe `--show-deprecated-hidden-metrics" ?

Comment (ehashman) [reply cmd/kube-apiserver/app/options/options.go]: But "deprecated" means something else. "Hidden" is well-defined by the metrics stability KEP, where the lifecycle of metrics is defined as:

Alpha > Stable > Deprecated > Hidden

Comment (brancz) [reply cmd/kube-apiserver/app/options/options.go]: Hidden is a state of a deprecated metric that is old enough no? At least that's how I always viewed it. This makes sense to me as well though.

Comment (liggitt) [reply cmd/kube-apiserver/app/options/options.go]: I'm also confused by this flag. Why didn't we stick with the original plan of allowing enabling a specific deprecated metric?

Comment (ehashman) [reply cmd/kube-apiserver/app/options/options.go]: That was the question I asked here on the KEP update: https://github.com/kubernetes/enhancements/pull/1358#discussion_r344640164

Comment (logicalhan) [reply cmd/kube-apiserver/app/options/options.go]: > Why didn't we stick with the original plan of allowing enabling a specific deprecated metric?

Because we want to reserve the ability to disable a metric (regardless of whether it is stable or not) at application boot and that functionality is orthogonal to metrics being auto-disabled by metrics stability framework.

Comment (logicalhan) [reply cmd/kube-apiserver/app/options/options.go]: The [original plan](https://github.com/kubernetes/enhancements/pull/946/files#diff-3d4ef632b2ce9fb3c84103dc166b833bR211) dictates providing a mechanism for turning off individual metrics via the command line. 

> ..all metrics should be able to be individually disabled by the cluster administrator, regardless of stability class. By default, all non-deprecated metrics will be automatically registered to the metrics endpoint unless explicitly blacklisted via a command line flag (i.e. '--disable-metrics=somebrokenmetric,anothermetric')

Comment (lavalamp) [reply cmd/kube-apiserver/app/options/options.go]: I suggest: `The previous version for which you want to show hidden metrics. Only the previous minor version is meaningful, other values will be ignored. The format is <major>.<minor>, e.g.: 1.16`

Comment (lavalamp) [reply cmd/kube-apiserver/app/options/options.go]: Given the discussion, I think it makes sense to add a comment here explaining the purpose of this (to avoid giving the admin a surprise at v N+2). This could also be said in the help text, e.g.: `The purpose of this format is make sure you have the opportunity to notice if the next release hides additional metrics, rather than being surprised when they are permanently removed in the release after that.`

**Comment by RainbowMango** [comment]
> I suggest: The previous version for which you want to show hidden metrics. Only the previous minor version is meaningful, other values will be ignored. The format is <major>.<minor>, e.g.: 1.16

New push with a minor change: `other values will be ignored` --> `other values will not be allowed` according to the logic of validation.

**Comment by lavalamp** [comment]
Does "not be allowed" mean apiserver won't start? I can't decide if that's good or not.

**Comment by liggitt** [comment]
silently ignoring an explicit option they asked for seems not good, especially for something like this where that would mean alerting/monitoring configurations could be making incorrect decisions based on absent metrics

File: staging/src/k8s.io/component-base/metrics/registry.go
Comment (lavalamp) [inline staging/src/k8s.io/component-base/metrics/registry.go]: s/not follows/does not follow/

File: staging/src/k8s.io/component-base/metrics/registry.go
Comment (lavalamp) [inline staging/src/k8s.io/component-base/metrics/registry.go]: I think you don't need any of this semver code. Only the previous minor version is allowed, so you can print it to a string and compare directly.

Comment (lavalamp) [reply staging/src/k8s.io/component-base/metrics/registry.go]: I think that can handle the above error condition, too. The error message can say, `--show-hidden-metrics-for-version must be omitted or have the value "%v". Only the previous minor version is allowed.`

**Comment by lavalamp** [comment]
OK, if @liggitt agrees then I'm sold, but that means the error handling code can be drastically simplified.

**Comment by RainbowMango** [comment]
> OK, if @liggitt agrees then I'm sold, but that means the error handling code can be drastically simplified.

Thanks, much clear now.

File: staging/src/k8s.io/component-base/metrics/registry.go
Comment (lavalamp) [inline staging/src/k8s.io/component-base/metrics/registry.go]: I don't think we're ever going to have a v2, but if we did, it would print 2.-1 which is not right :)

Feel free to fix in a follow-up if you want.

**Pull request merged**
